src/generic/fetch.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Contains utility methods using 'fetch'.
+ */
+
+import {ApiError, ErrorCode} from '../core/api_error';
+import {BFSCallback} from '../core/file_system';
+
+/**
+ * Asynchronously download a file as a buffer or a JSON object.
+ * Note that the third function signature with a non-specialized type is
+ * invalid, but TypeScript requires it when you specialize string arguments to
+ * constants.
+ * @hidden
+ */
+export function fetchFileAsync(p: string, type: 'buffer', cb: BFSCallback<Buffer>): void;
+export function fetchFileAsync(p: string, type: 'json', cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void {
+  fetch(p)
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch error."));
+      } else {
+        switch (type) {
+          case 'buffer':
+            res.arrayBuffer()
+              .then((buf) => cb(null, Buffer.from(buf)))
+              .catch((err) => cb(new ApiError(0, err.message)));
+            break;
+          case 'json':
+            res.json()
+              .then((json) => cb(null, json))
+              .catch((err) => cb(new ApiError(0, err.message)));
+            break;
+          default:
+            throw new ApiError(ErrorCode.EINVAL, "Invalid download type: " + type);
+        }
+      }
+    })
+    .catch((err) => cb(new ApiError(0, err.message)));
```

Comment:
As usual, I'm slightly wrong! There is one edge case: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Exceptions

> TypeError: Since Firefox 43, fetch() will throw a TypeError if the URL has credentials, such as http://user:password@example.com.

Obviously that's an error that we'd want to bubble up to the user somehow. Should we wrap the whole thing in a try/catch and pass the error to the callback?